File: 501.二叉搜索树中的众数.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> maxCntNums;
    int cur = 1;
    int max = 1;
    TreeNode* pre = NULL;

    vector<int> findMode(TreeNode* root) {
        inOrder(root);
        return maxCntNums;
    }

    void inOrder(TreeNode* root){
        if(!root)   return;
        inOrder(root->left);
        if(pre){
            if(pre->val == root->val)
                ++cur;
            else
                cur = 1;
        }
        if(cur > max) {
            max = cur;
            // 清空数组 
            maxCntNums.clear();
            maxCntNums.push_back(root->val);
        } else if (cur == max) {
            maxCntNums.push_back(root->val);
        }
        pre = root;
        inOrder(root->right);
    }
};

File: 二叉树/501.二叉搜索树中的众数.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
 var findMode = function (root) {
    let res = [];
    let pre = null;
    let cur = 1;
    let max = 0;

    const inOrder = (root) => {
        if (root === null)
            return;
        inOrder(root.left);
        if (pre !== null) {
            if (pre.val === root.val)
                ++cur;
            else
                cur = 1;
        }
        if (cur > max) {
            max = cur;
            res = [];
            res.push(root.val);
        } else if (cur === max) {
            res.push(root.val);
        }
        pre = root;
        inOrder(root.right);
    }

    inOrder(root);
    return res;
};